### ISSUE TITLE
NaN in CPU training of resnet

### ISSUE THREAD

**Issue by sfraczek** [PaddlePaddle/Paddle#26673]
NaNs in training Resnet on CPU from models repo (without MKLDNN)

### Paddle models version
commit 64cde5d16de78181e348abc85fa45cee29512ec6
Author: LiuHao <liuhao19900412@gmail.com>
Date:   Thu Aug 6 18:23:44 2020 +0800

    Update run_ernie_classifier.py (#4790)

### Paddle version
I chcked below two commits and some in between and all don't work
#### This I chcked first
commit 36868e840c582047480f4c999ddf1f5f004de678
Author: yukavio <67678385+yukavio@users.noreply.github.com>
Date:   Mon Aug 24 18:35:03 2020 +0800

    fix one_hot example doc test=document_fix (#26585)

    * fix one_hot example doc test=document_fix

    * fix some bug

#### This I checked later too
commit 9d2bd0ac387080c5a047565ba9f28d380d1c42f1
Author: 123malin <malin10@baidu.com>
Date:   Wed Jun 3 14:05:17 2020 +0800

    downpour_worker增加try_catch机制，打印program所有参数 (#24700)

    * test=develop, add try_catch for debug

### Other info
Build type: Release
CPU: Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
OS: Ubuntu 16.04.5 LTS
Python: 2.7.12

### To reproduce
in `models/dygraph/resnet`
Command: `python train.py`
change place from CUDA to CPU
```diff
-    place = fluid.CUDAPlace(fluid.dygraph.parallel.Env().dev_id) \
-        if args.use_data_parallel else fluid.CUDAPlace(0)
+    place = fluid.CPUPlace()
```

### Log
```
[Epoch 0, batch 0] loss 4.60813, acc1 0.03125, acc5 0.06250, batch_cost: 14.17363 s, reader_cost: 0.14763 s
[Epoch 0, batch 10] loss nan, acc1 0.01136, acc5 0.06250, batch_cost: 11.84618 s, reader_cost: 0.00004 s
```

**Comment by phlrain** [comment]
I try to reproduce result, but I don't get NAN. 

Log.
```
[Epoch 0, batch 0] loss 4.83761, acc1 0.03125, acc5 0.06250, batch_cost: 28.49166 s, reader_cost: 0.65109 s
[Epoch 0, batch 10] loss 13.95101, acc1 0.02841, acc5 0.11648, batch_cost: 25.39008 s, reader_cost: 0.00007 s
```

From my experience, the loss in batch 10 is around 11-14, becaure of data shuffle and weight initialization.

Cound you please me some more information.

Like cmake command (cmake ../ -DWITH_GPU=OFF ...) and docker information if used.

**Comment by arlesniak** [comment]
The training has been run on bare metal Xeon.

cmake .. -DWITH_GPU=OFF -DWITH_DISTRIBUTE=OFF -DWITH_MKLDNN=ON -DCMAKE_EXPORT_COMPILE_COMMANDS=OFF -DWITH_PROFILER=OFF -DON_INFER=ON -DCMAKE_BUILD_TYPE=Release -DWITH_TESTING=ON -DWITH_INFERENCE_API_TEST=ON -DWITH_NCCL=OFF -DWITH_COVERAGE=OFF

**Paddle repo (branch develop):**
newest commit id: 6947a58a1f52bed347341d373ce86452b17c3366 Mon Sep 14 20:14:26 2020 

**Model repo (branch develop):**
newest commit id: bde994e12c93b4f1c6f0b195653192a76a6015ab  Tue Sep 15 15:13:27

**cmd line:**
models/dygraph/resnet$ python train.py --use_gpu=0 --ce

**result:**

> ce mode
> [Epoch 0, batch 0] loss 4.82573, acc1 0.00000, acc5 0.00000, batch_cost: 11.09144 s, reader_cost: 0.18639 s
> [Epoch 0, batch 10] loss nan, acc1 0.00852, acc5 0.03693, batch_cost: 9.12530 s, reader_cost: 0.00004 s

**cmd line:**
models/dygraph/resnet$ python train.py --use_gpu=0

**result:**

> [Epoch 0, batch 0] loss 4.76966, acc1 0.00000, acc5 0.00000, batch_cost: 11.87695 s, reader_cost: 0.18596 s
> [Epoch 0, batch 10] loss nan, acc1 0.00852, acc5 0.04830, batch_cost: 8.95958 s, reader_cost: 0.00004 s

**Comment by lidanqing-vv** [comment]
Hi, @arlesniak What is this machine ? Is it 6248? use `lscpu` and update to Baidu
And also if possible try to reproduyce on 6148, because baidu can not reproduce this error...

**Comment by arlesniak** [comment]
The above result are from :
Model name:            Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz

**Comment by arlesniak** [comment]
****************************************
Paddle version: 6947a58a1f52bed347341d373ce86452b17c3366 Date:   Mon Sep 14 20:14:26 2020 +0800
Paddle branch: develop
Paddle With CUDA: False

Paddle models repo: https://github.com/PaddlePaddle/models.git
Paddle models version: bde994e12c93b4f1c6f0b195653192a76a6015ab Tue Sep 15 15:13:27 2020 +0800
Paddle models branch: develop

CPU: Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz

OS: Ubuntu 16.04.6 LTS
Python version: 2.7.12

CUDA version: None
cuDNN version: None.None.None
Nvidia driver version: None

Cmake orders: -DWITH_GPU=OFF -DWITH_DISTRIBUTE=OFF -DWITH_MKLDNN=ON -DCMAKE_EXPORT_COMPILE_COMMANDS=OFF -DWITH_PROFILER=OFF -DON_INFER=ON -DCMAKE_BUILD_TYPE=Release -DWITH_TESTING=ON -DWITH_INFERENCE_API_TEST=ON -DWITH_NCCL=OFF -DWITH_COVERAGE=OFF

-C++ version : gcc (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
-cmake version: cmake version 3.17.3
****************************************
Steps to reproduce:
having Paddle built and models repo cloned run the following cmd (please use your paths):

>  PYTHONPATH=Paddle/build/python/ python models/dygraph/resnet/train.py --use_gpu=0 --ce

**Comment by arlesniak** [comment]
@phlrain @luotao1  Could you reproduce the issue with the updated reproducing details. please ?

**Comment by phlrain** [comment]
> @phlrain @luotao1 Could you reproduce the issue with the updated reproducing details. please ?

I can get an NAN by the updated reproducing details. I will check the reason why cause this.

**Comment by luotao1** [comment]
@phlrain Could you give the environment and cmake of https://github.com/PaddlePaddle/Paddle/issues/26673#issuecomment-692072288 ?

**Comment by arlesniak** [comment]
@phlrain @luotao1 Do you have any update of status on the issue ?

**Comment by lidanqing-vv** [comment]
@phlrain @luotao1  Hi, Please help fixing this issue. Since we already enbaled dyGraph oneDNN resnet50 training, and we also know for mobilenetv1, and v2, oneDNN training has 5-6X speed-up compared to Native config version, but for resnet50, since NativeConfig verision has NAN error, we can not calculate the speed-up.

**Comment by luotao1** [comment]
@arlesniak @lidanqing-intel We find the reason, the cmake option`-DON_INFER=ON` is for inference, you should set `-DON_INFER=OFF` which is for training. In this cmake option, CPU training of resnet is normal.
You can have a double-check.

Maybe the reason is that softmax has different version on Inference and Training mode (see #14337), and some other differences.

**Comment by arlesniak** [comment]
Thank you for the info we will check internally.

**Comment by lidanqing-vv** [comment]
We have verified and it solved the problem.
